### PR TITLE
feat: app billing info

### PR DESCRIPTION
**Pull request by Noums26**

### CODE REVIEW

File: src/operations/billing/BillingInformation.tsx [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import {EmailField, Show, SimpleShowLayout, TextField} from "react-admin";
+
```
Comment (YumeT023) [inline src/operations/billing/BillingInformation.tsx:2]: rename the file into the component's name

File: src/providers/paymentDetailsProvider.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import {PaymentCustomer} from "@jcloudify-api/typescript-client";
+import {PojaDataProvider, ToRecord} from "./types";
+import {authProvider} from "./authProvider";
+import {paymentApi, unwrap} from "../services/poja-api";
```
Comment (YumeT023) [inline src/providers/paymentDetailsProvider.ts:4]: fix imports path

File: src/operations/apps/AppBilling.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      title=" "
+      resource="billingInfo"
+      id={env.id}
+      queryOptions={{meta: {appId, targetResource: "environment"}}}
+    >
+      <SimpleShowLayout>
+        <FunctionField
+          render={(resource: BillingInfo) => (
+            <Typography variant="body2">
+              {env.environment_type} - $ {resource.computed_price}
```
Comment (YumeT023) [inline src/operations/apps/AppBilling.tsx:73]: use EnvironmentType component to display component environment_type

File: src/operations/apps/appShowViews.tsx [hunk 4]
```diff
@@ -20,4 +21,8 @@ export const appShowViews: RouteMap = {
     "$$index": <AppDeploymentList />,
     ":deploymentId": <AppDeploymentShow />,
   },
+  billing: {
+    "$$index": <AppBilling />,
+    ":billingId": <></>,
```
Comment (YumeT023) [inline src/operations/apps/appShowViews.tsx:26]: useless

File: src/operations/billing/Billing.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import {Box, Card, CardContent, Stack, Typography} from "@mui/material";
+import {ShowBillingInfo, ShowBillingInformationSummary} from "./";
+import {authProvider} from "@/providers";
+
+export const BillingSummary: React.FC = () => {
+  const customerId = authProvider.getCachedWhoami()?.user?.stripe_id!;
```
Comment (YumeT023) [inline src/operations/billing/Billing.tsx:6]: this can be retrieved from the provider !

File: src/operations/billing/Payments.tsx [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import {FC} from "react";
+import {Box, Stack} from "@mui/material";
+import {BillingSummary} from "./Billing";
```
Comment (YumeT023) [inline src/operations/billing/Payments.tsx:3]: fix import path

File: src/operations/billing/Payments.tsx [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+export const Payments: FC = () => {
+  return (
```
Comment (YumeT023) [inline src/operations/billing/Payments.tsx:6]: inline return

File: src/operations/billing/ShowBillingInfo.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+}> = ({targetId = "summary", targetResource = "summary"}) => {
+  return (
+    <Show
+      title=" "
+      emptyWhileLoading
+      resource="billingInfo"
+      id={targetId}
+      queryOptions={{meta: {targetResource}}}
+    >
+      <SimpleShowLayout>
```
Comment (YumeT023) [inline src/operations/billing/ShowBillingInfo.tsx:23]: no!

use the custom ShowLayout from operations/components/show

remove emptyWhileLoading

File: src/providers/billingInfoProvider.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        )) as ToRecord<BillingInfo>;
+        return {...envBillingInfo, id: targetId};
+      default:
+        const billingInfo = (await unwrap(() =>
+          billingApi().getUserBillingInfo(uid, startDate, currentDate)
+        )) as ToRecord<BillingInfo>;
+        return {...billingInfo, id: targetId};
+    }
+  },
+  getList: function (): Promise<PagedResponse<ToRecord<BillingInfo>>> {
```
Comment (YumeT023) [inline src/providers/billingInfoProvider.ts:45]: rm return types

File: src/providers/billingInfoProvider.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  delete: function (): Promise<ToRecord<BillingInfo>> {
+    throw new Error("Function not implemented.");
+  },
+  deleteMany: function (): Promise<ToRecord<BillingInfo>[]> {
+    throw new Error("Function not implemented.");
+  },
+};
+
+const mapToBillingInfo = (
+  data: BillingInfo[],
```
Comment (YumeT023) [inline src/providers/billingInfoProvider.ts:63]: data -> billingInfos

File: src/providers/billingInfoProvider.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    const currentDate = new Date();
+    const startDate = new Date(
+      currentDate.getFullYear(),
+      currentDate.getMonth(),
+      1
+    );
+
+    switch (meta.targetResource) {
+      case "application":
+        const res = await unwrap(() =>
```
Comment (YumeT023) [inline src/providers/billingInfoProvider.ts:18]: res -> userAppBillingInfo

File: src/operations/billing/ShowBillingInfo.tsx [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        <Labeled>
+          <FunctionField
+            label="Current month"
+            render={(resource: BillingInfo) => (
+              <Typography variant={"h4"}>
```
Comment (YumeT023) [inline src/operations/billing/ShowBillingInfo.tsx:28]: no need explicit typing in the render fn, 

instead


<FunctionField<ToRecord<BillingInfo>> 
  render={...}
/>

**Review by YumeT023 — CHANGES_REQUESTED** [review]
filename -> main component name
distinguish the payment from the billing itself!

File: src/operations/apps/AppBilling.tsx [hunk 13]
```diff
@@ -38,20 +41,15 @@ const BillingShow = () => {
 const AppBillingDetails: React.FC<{appId: string}> = ({appId}) => {
   const {data: environments = []} = useListContext<ToRecord<Environment>>();
```
Comment (YumeT023) [inline src/operations/apps/AppBilling.tsx:42]: use, WithListContext component to get the data prop to map

File: src/operations/billing/Billing.tsx [hunk 14]
```diff
@@ -1,30 +1,27 @@
-import {Box, Card, CardContent, Stack, Typography} from "@mui/material";
-import {ShowBillingInfo, ShowBillingInformationSummary} from "./";
-import {authProvider} from "@/providers";
+import {Box, Stack} from "@mui/material";
+import {ContainerWithHeading} from "@/components/container";
+import {PaymentDeatilsShow} from "@/operations/payments";
+import {ShowBillingInfo} from "./";
```
Comment (YumeT023) [inline src/operations/billing/Billing.tsx:4]: fix import path

File: src/operations/billing/Billing.tsx [hunk 15]
```diff
@@ -1,30 +1,27 @@
...
-    </Card>
-  );
-};
+export const Billing: React.FC = () => (
+  <Stack direction="column" spacing={1}>
+    <BillingSummary />
+  </Stack>
+);
+
+export const BillingSummary: React.FC = () => (
```
Comment (YumeT023) [inline src/operations/billing/Billing.tsx:12]: do not export

File: src/operations/billing/Billing.tsx [hunk 16]
```diff
@@ -1,30 +1,27 @@
...
+    <Stack
+      direction={{sm: "column", md: "row"}}
+      spacing={1}
+      alignItems="stretch"
+    >
+      <Box sx={{width: "100%"}}>
+        <ShowBillingInfo />
+      </Box>
+      <Box sx={{width: "100%"}}>
+        <PaymentDeatilsShow />
```
Comment (YumeT023) [inline src/operations/billing/Billing.tsx:23]: PaymentDetailsShow typo

File: src/operations/billing/Billing.tsx (hunk 15 above)
Comment (YumeT023) [inline src/operations/billing/Billing.tsx:12]: why do you separate the BillingSummary and the Billing component ?... ?

File: src/operations/billing/ShowBillingInfo.tsx [hunk 17]
```diff
@@ -1,43 +1,39 @@
...
-        </Labeled>
-      </SimpleShowLayout>
-    </Show>
+      <ShowLayout>
+        <Stack direction="column" spacing={1}>
+          <Labeled>
+            <FunctionField<ToRecord<BillingInfo>>
+              label="Current month"
+              render={(record) => (
+                <Typography variant={"h4"}>
```
Comment (YumeT023) [inline src/operations/billing/ShowBillingInfo.tsx:23]: no need curly brackets

File: src/operations/billing/index.ts [hunk 18]
```diff
@@ -1,16 +1,15 @@
 import {Payment as PaymentIcon} from "@mui/icons-material";
 import {Operation} from "@/operations/types";
-import {Payments} from "./Payments";
+import {Billing} from "./Billing";
 
 export const billing: Operation = {
   name: "billing",
-  list: Payments,
+  list: Billing,
```
Comment (YumeT023) [inline src/operations/billing/index.ts:7]: put // TODO: custom Menu

File: src/operations/billing/index.ts [hunk 19]
```diff
@@ -1,16 +1,15 @@
...
   icon: PaymentIcon,
   options: {label: "Billing"},
 };
 
-export * from "./Payments";
+export * from "../payments/Payments";
```
Comment (YumeT023) [inline src/operations/billing/index.ts:12]: ??!!

File: src/operations/payments/Payments.tsx [hunk 20]
```diff
@@ -0,0 +1,8 @@
+import {Stack} from "@mui/material";
+import {BillingSummary} from "@/operations/billing";
+
+export const Payments: React.FC = () => (
+  <Stack direction="column" spacing={1}>
+    <BillingSummary />
```
Comment (YumeT023) [inline src/operations/payments/Payments.tsx:6]: separate the Billing itself and the BillingSummary

File: src/providers/billingInfoProvider.ts [hunk 21]
```diff
@@ -1,10 +1,10 @@
 import {BillingInfo} from "@jcloudify-api/typescript-client";
-import {PagedResponse, PojaDataProvider, ToRecord} from "./types";
 import {billingApi, unwrap} from "@/services/poja-api";
 import {authProvider} from "./authProvider";
+import {PojaDataProvider, ToRecord} from "./types";
```
Comment (YumeT023) [inline src/providers/billingInfoProvider.ts:4]: fix imports path

File: src/providers/billingInfoProvider.ts [hunk 22]
```diff
@@ -1,10 +1,10 @@
 import {BillingInfo} from "@jcloudify-api/typescript-client";
-import {PagedResponse, PojaDataProvider, ToRecord} from "./types";
 import {billingApi, unwrap} from "@/services/poja-api";
 import {authProvider} from "./authProvider";
```
Comment (YumeT023) [inline src/providers/billingInfoProvider.ts:3]: import paths

File: src/operations/apps/AppBilling.tsx [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+import {WithTab} from "@/components/tab";
+import {ContainerWithHeading} from "@/components/container";
+
+export const AppBilling: React.FC = () => (
+  <WithTab tab="Billing">
+    <BillingShow />
+  </WithTab>
+);
+
+const BillingShow = () => {
```
Comment (YumeT023) [inline src/operations/apps/AppBilling.tsx:24]: put this in ./billing/BillingShow.tsx

File: src/operations/apps/AppBilling.tsx [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+          <BillingInfoDetails env={env} appId={appId} />
+        ))
+      ) : (
+        <Typography variant="body2">No results found</Typography>
+      )}
+    </ContainerWithHeading>
+  );
+};
+
+const BillingInfoDetails: React.FC<{env: Environment; appId: string}> = ({
```
Comment (YumeT023) [inline src/operations/apps/AppBilling.tsx:56]: amboary pr

**Review by YumeT023 — CHANGES_REQUESTED** [review]

File: src/operations/billing/Billing.tsx [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+
+const BillingSummary: React.FC = () => (
+  <ContainerWithHeading title="Payment Details Summary">
+    <Stack
+      direction={{sm: "column", md: "row"}}
+      spacing={1}
+      alignItems="stretch"
+    >
+      <Box sx={{width: "100%"}}>
+        <ShowBillingInfo />
```
Comment (Noums26) [inline src/operations/billing/Billing.tsx:20]: BillingInfoShow

File: src/operations/billing/BillingShow.tsx [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+          ) : (
+            <Typography variant="body2">No results found</Typography>
+          )}
+        </ContainerWithHeading>
+      )}
+    />
+  );
+};
+
+const BillingInfoDetails: React.FC<{env: Environment; appId: string}> = ({
```
Comment (Noums26) [inline src/operations/billing/BillingShow.tsx:51]: BillingInfoDetailsShow

File: src/operations/billing/BillingInfoShow.tsx [hunk 27]
```diff
@@ -0,0 +1,37 @@
+import {BillingInfo} from "@jcloudify-api/typescript-client";
+import {DateField, FunctionField, Labeled, ShowBase} from "react-admin";
+import {Stack, Typography} from "@mui/material";
+import {ToRecord} from "@/providers";
+import {ShowLayout} from "@/operations/components/show";
+
+export const BillingInfoShow: React.FC<{
+  targetId?: string;
+  targetResource?: string;
+}> = ({targetId = "summary", targetResource = "summary"}) => (
```
Comment (YumeT023) [inline src/operations/billing/BillingInfoShow.tsx:10]: "*"

File: src/operations/billing/BillingShow.tsx [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+  WithListContext,
+} from "react-admin";
+import {useParams} from "react-router-dom";
+import {Box, Stack, Typography} from "@mui/material";
+import {ToRecord} from "@/providers";
+import {BillingInfoShow} from "@/operations/billing";
+import {ShowLayout} from "@/operations/components/show";
+import {SimpleListEmpty} from "@/operations/components/list";
+import {EnvironmentType} from "@/operations/environments";
+import {ContainerWithHeading} from "@/components/container";
```
Comment (YumeT023) [inline src/operations/billing/BillingShow.tsx:16]: @/components goes before @/operations/...

@/operations/.... goes before @/providers

**Review by YumeT023 — CHANGES_REQUESTED** [review]

File: src/operations/billing/BillingShow.tsx [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+import {ShowLayout} from "@/operations/components/show";
+import {SimpleListEmpty} from "@/operations/components/list";
+import {BillingInfoShow} from "@/operations/billing";
+import {EnvironmentType} from "@/operations/environments";
+import {ToRecord} from "@/providers";
+import {colors} from "@/themes";
+
+export const BillingShow: React.FC<{appId: string}> = ({appId}) => {
+  return (
```
Comment (YumeT023) [inline src/operations/billing/BillingShow.tsx:19]: inline return

**Review by YumeT023 — CHANGES_REQUESTED** [review]

**Pull request merged**